.github/copilot-instructions.md

```diff
@@ -0,0 +1,211 @@
+   - cargo test --verbose  
+   - cargo clippy -- -D warnings
+   - cargo fmt -- --check
+   - Security audit via rustsec/audit-check
+
+2. **release.yml**: Semantic release workflow
+   - Triggered manually
+   - Uses semantic-release with conventional commits
+
+3. **docs.yml**: Quarto documentation publishing to GitHub Pages
+
+### Pre-commit Validation
+The CI exactly replicates these commands - ensure all pass locally:
+```bash
+cargo build --verbose
+cargo test --verbose
+cargo clippy -- -D warnings
+cargo fmt -- --check
+```
+
+### Golden Test System
+The project uses snapshot testing via `tests/golden_cases.rs`:
+- Each `tests/cases/*` directory contains `input.qmd` and `expected.qmd`
+- Tests verify formatting is idempotent (format twice = format once)
+- Use `UPDATE_EXPECTED=1 cargo test` to update expected outputs (BE CAREFUL)
+
+## Key Development Facts
+
+### Dependencies
+- **clap**: CLI argument parsing with derive macros
+- **rowan**: Red-green tree for CST (Concrete Syntax Tree)  
+- **regex**: Pattern matching for lexing
+- **textwrap**: Text wrapping functionality
+- **toml**: Configuration file parsing
+- **serde**: Serialization for config structs
+
+### Testing Approach  
+- Unit tests embedded in source modules (34 tests)
+- Integration tests in `tests/format/` (39 tests) 
+- Golden tests comparing input/expected pairs (1 comprehensive test covering 14 scenarios)
+- Property: formatting is idempotent
+
+### Formatting Rules
+- Default 80 character line width
+- Preserves frontmatter and code blocks
+- Converts setext headings to ATX format
+- Handles Quarto-specific syntax (fenced divs, math blocks)
+- Wraps paragraphs but preserves inline code/math whitespace
+
+## Configuration Files and Settings
+- `.quartofmt.toml`: Project-specific config (line_width, line-ending, wrap mode)
+- `.envrc`: direnv configuration for Nix environment
+- `.gitignore`: Excludes target/, devenv artifacts, Nix build outputs
+- `devenv.nix`: Development environment with go-task, quarto, wasm-pack
+
+## Web Playground
+The `docs/playground/` contains a WASM-based web interface:
+- Built via `wasm-pack build --target web` in `crates/quartofmt-wasm/`
+- Uses TypeScript bindings for browser integration
+- Served via local Python HTTP server for development
+
+## Important Notes for Code Changes
+
+### DO:
+- Run full test suite after every change: `cargo test` 
+- Ensure clippy passes: `cargo clippy -- -D warnings`
+- Test CLI functionality after building release binary
+- Consider idempotency - formatting twice should equal formatting once
+- Update golden test expectations carefully with `UPDATE_EXPECTED=1 cargo test`
+
+### DON'T:
+- Assume task runner is available - use direct cargo commands
+- Break the hierarchical config system (explicit > local > XDG > default)  
+- Change core formatting without extensive golden test verification
+- Ignore the 2 build warnings (they're acceptable dead code)
+
+### Architecture Dependencies
```

Comment:
I intend to move away from this structure to a block parser -> inline parser structure. The block parser should capture block structs, including nested ones. The inline parser/lexer will then run on these to mark up the inline syntax